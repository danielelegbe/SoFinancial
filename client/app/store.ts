import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../features/user/userSlice';
import chatReducer from '../features/chat/chatSlice';

export const store = configureStore({
  reducer: {
    user: userReducer,
    chat: chatReducer,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
