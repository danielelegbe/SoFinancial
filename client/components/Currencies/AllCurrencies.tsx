import { CircularProgress, Flex, HStack } from '@chakra-ui/react';
import axios from 'axios';
import React, { useEffect, useState } from 'react';
import io from 'socket.io-client';
import Currency from './Currency';

export interface ICurrency {
  s?: string;
  p?: number;
}

const Currencies = () => {
  const [BTC, setBTC] = useState<ICurrency | null>(null);
  const [GBPUSD, setGBPUSD] = useState<ICurrency | null>(null);
  const [EURUSD, setEURUSD] = useState<ICurrency | null>(null);
  const [USDCAD, setUSDCAD] = useState<ICurrency | null>(null);

  useEffect(() => {
    const socket = io();
    axios('/api/currencies').then(() => {
      io('/');
    });
    socket.on('currencies', (data) => {
      const currencies = JSON.parse(data);

      if (currencies?.data) {
        const BTC: ICurrency = currencies.data.find(
          (currency: ICurrency) => currency.s === 'BINANCE:BTCUSDT'
        );
        const GBPUSD: ICurrency = currencies.data.find(
          (currency: ICurrency) => currency.s === 'OANDA:GBP_USD'
        );
        const EURUSD: ICurrency = currencies.data.find(
          (currency: ICurrency) => currency.s === 'OANDA:EUR_USD'
        );
        const USDCAD: ICurrency = currencies.data.find(
          (currency: ICurrency) => currency.s === 'OANDA:USD_CAD'
        );
        BTC && setBTC(BTC);
        GBPUSD && setGBPUSD(GBPUSD);
        EURUSD && setEURUSD(EURUSD);
        USDCAD && setUSDCAD(USDCAD);
      }
    });

    return () => {
      socket.disconnect();
    };
  }, []);

  return (
    <Flex justify="center" boxShadow="md" py={2} align="center" bg="gray.100">
      {!BTC && <CircularProgress isIndeterminate color="blue.500" />}
      <HStack spacing={5}>
        {BTC && <Currency {...BTC} />}
        {GBPUSD && <Currency {...GBPUSD} />}
        {EURUSD && <Currency {...EURUSD} />}
        {USDCAD && <Currency {...USDCAD} />}
      </HStack>
    </Flex>
  );
};

export default Currencies;
