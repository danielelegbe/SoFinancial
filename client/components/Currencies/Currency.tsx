import { MinusIcon, TriangleDownIcon, TriangleUpIcon } from '@chakra-ui/icons';
import { ComponentWithAs, HStack, IconProps, Text } from '@chakra-ui/react';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../app/store';
import { setLastPrice } from '../../features/currencies/currenciesSlice';
import { ICurrency } from './AllCurrencies';

const Currency = ({ s: symbol, p: price }: ICurrency) => {
  const [textColor, setTextColor] = useState('');
  const [icon, setIcon] = useState(<MinusIcon />);
  const dispatch = useDispatch();

  const { p: lastPrice } = useSelector((state: RootState) => state.currency);

  useEffect(() => {
    if (!lastPrice || lastPrice === price) {
      setTextColor('black');
      setIcon(<MinusIcon boxSize={3} />);
    } else {
      if (price && price > lastPrice) {
        setTextColor('green');
        setIcon(<TriangleUpIcon color="green" boxSize={3} />);
      } else {
        setTextColor('red');
        setIcon(<TriangleDownIcon color="red" boxSize={3} />);
      }
    }
    price && dispatch(setLastPrice(price));
  }, [price]);

  return (
    <HStack
      border="1px"
      borderColor="gray.200"
      rounded="md"
      boxShadow="md"
      p={3}
      justify="space-between"
    >
      <Text fontSize="sm" fontWeight="bold">
        {symbol}
      </Text>
      <Text fontSize="sm" color={textColor}>
        {price && parseFloat(price.toFixed(5))}
      </Text>
      {icon}
    </HStack>
  );
};

export default Currency;
