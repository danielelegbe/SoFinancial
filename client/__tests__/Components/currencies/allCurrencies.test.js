import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Currencies from '../../../components/Currencies/AllCurrencies';
import { Provider } from 'react-redux';
import { store } from '../../../app/store';
import { beforeEach } from 'jest-circus';

afterEach(() => {
  cleanup();
})
beforeEach(() => {
})

test('test', () => {
  render(<Provider store={store}><Currencies /></Provider>)
  const allCurr = screen.getByTestId('all-currencies')
})

test('snapshot test', () => {
  const currencyScreen = renderer.create(<Provider store={store}><Currencies /></Provider>).toJSON();
  expect(currencyScreen).toMatchSnapshot()
})