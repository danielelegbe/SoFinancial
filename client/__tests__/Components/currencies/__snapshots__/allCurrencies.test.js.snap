// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="css-io698s"
  data-testid="all-currencies"
>
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    className="chakra-progress css-120wkjd"
    data-indeterminate=""
    role="progressbar"
  >
    <svg
      className="css-jf70f3"
      viewBox="0 0 100 100"
    >
      <circle
        className="chakra-progress__track css-1bhxzgy"
        cx={50}
        cy={50}
        r={42}
        strokeWidth="10px"
      />
      <circle
        className="chakra-progress__indicator css-p6i8i8"
        cx={50}
        cy={50}
        r={42}
        strokeWidth="10px"
      />
    </svg>
  </div>
  <div
    className="chakra-stack css-15lgeks"
  />
</div>
`;
